### PR TITLE
Python 3.12

### PR DESCRIPTION
**Pull request by dbrnz**
Here it is!

### CODE REVIEW

**Review by agarny — COMMENTED** [review]
Thanks for this @dbrnz. It all compiles fine. However, besides my cleaning up and the few comments that I have made, I am having some issues with the Python scripts (I have only tested on macOS). For instance (note commit 44527774e to prevent the debug warning):

- `ipython`: seems to be fine;
- `jupyter`: seems to be fine;
- `jupyterconsole`: I get various tracebacks (see [log](https://github.com/opencor/opencor/files/14848959/jupyterconsole.log));
- `jupyterlab`: although it fires up JupyerLab fine and I could open the [lorenz.ipynb](https://github.com/opencor/opencor/blob/master/models/tests/jupyter/lorenz.ipynb) file, I got various tracebacks (see [log](https://github.com/opencor/opencor/files/14848962/jupyterlab.log));
- `jupyternotebook`: same as `jupyterlab` (see [log](https://github.com/opencor/opencor/files/14848958/jupyernotebook.log)); and
- `pythonshell`: seems to be fine except for the fact that I couldn't use the `Up` and `Down` keys to recall previous commands. Instead, I got `^[[A` and `^[[B`, respectively (see [log](https://github.com/opencor/opencor/files/14848963/pythonshell.log)).

Otherwise, as you might have seen, only the Windows debug CI passes (even though there are no new Windows debug binaries?!). All the other ones fail for some reasons.

File: src/plugins/thirdParty/PythonPackages/CMakeLists.txt [hunk 1]
```diff
@@ -122,65 +141,46 @@ else()
...
+        INSTALL_COMMAND
+            ${CMAKE_COMMAND} -E env "PYTHONHOME=${PYTHON_ROOT_DIR}"
+                                env "PYTHONPATH=${FULL_LOCAL_SITE_PACKAGES_DIR}"
+                                env "LD_LIBRARY_PATH=${BUILT_LIBRARIES_DIR}"
+                                ${PYTHON_EXECUTABLE} -s ${PIP_SOURCE_DIR}/get-pip.py --prefix ${FULL_LOCAL_EXTERNAL_PACKAGE_DIR}
+                                                                                     --no-wheel
+        STEP_TARGETS install
+        LOG_DIR
+            "${CMAKE_SOURCE_DIR}/log"
+        LOG_INSTALL TRUE
```
[the remark this answers is not in the document]
Comment (dbrnz) [reply src/plugins/thirdParty/PythonPackages/CMakeLists.txt]: That's fine -- leftover from me trying to understand why things didn't always work.

File: src/plugins/thirdParty/Python/CMakeLists.txt [hunk 2]
```diff
@@ -442,9 +518,14 @@ else()
             -DBUILD_TESTING=OFF
             -DINSTALL_TEST=OFF
 
-            # Optionally patch sources (on Windows)
+            # We don't want the Window's distutils installer
+            # (which in fact doesn't build for Python 3.10)
```
[the remark this answers is not in the document]
Comment (dbrnz) [reply src/plugins/thirdParty/Python/CMakeLists.txt:514]: It turns out that `distutils` was deprecated in Python 3.10 and removed in 3.12.

File: src/plugins/thirdParty/Python/CMakeLists.txt [hunk 3]
```diff
@@ -400,14 +473,17 @@ else()
...
+
     string(REPLACE ";" "|"
            PYTHON_OPENSSL_LIBRARIES "${OPENSSL_LIBRARIES}")
 
     ExternalProject_Add(${PACKAGE_BUILD}
         INSTALL_DIR
             ${FULL_LOCAL_EXTERNAL_PACKAGE_DIR}
         URL
-            https://github.com/opencor/${PACKAGE_REPOSITORY}-cmake-buildsystem/archive/refs/tags/${GIT_TAG}.tar.gz
+            https://github.com/opencor/portable-python-cmake-buildsystem/archive/opencor-python.3.12.tar.gz
```
[the remark this answers is not in the document]
Comment (dbrnz) [reply src/plugins/thirdParty/Python/CMakeLists.txt]: Fixed in commit [5406c32](https://github.com/opencor/opencor/pull/2773/commits/5406c321774eb18a79a4396c4d88afd64c441b6d)

File: src/plugins/thirdParty/PythonPackages/CMakeLists.txt [hunk 4]
```diff
@@ -122,65 +141,46 @@ else()
...
+            ""
+        BUILD_COMMAND
+            ""
+        INSTALL_COMMAND
+            ${CMAKE_COMMAND} -E env "PYTHONHOME=${PYTHON_ROOT_DIR}"
+                                env "PYTHONPATH=${FULL_LOCAL_SITE_PACKAGES_DIR}"
+                                env "LD_LIBRARY_PATH=${BUILT_LIBRARIES_DIR}"
+                                ${PYTHON_EXECUTABLE} -s ${PIP_SOURCE_DIR}/get-pip.py --prefix ${FULL_LOCAL_EXTERNAL_PACKAGE_DIR}
+                                                                                     --no-wheel
+        STEP_TARGETS install
```
[the remark this answers is not in the document]
Comment (dbrnz) [reply src/plugins/thirdParty/PythonPackages/CMakeLists.txt]: Yes, this is how we install `pip`.

**Comment by dbrnz** [comment]
Thanks for your testing and updates @agarny. You should have spotted my comments above. Of the other items:

> jupyterconsole: I get various tracebacks (see [log](https://github.com/opencor/opencor/files/14848959/jupyterconsole.log));
> jupyterlab: although it fires up JupyerLab fine and I could open the [lorenz.ipynb](https://github.com/opencor/opencor/blob/master/models/tests/jupyter/lorenz.ipynb) file, I got various tracebacks (see [log](https://github.com/opencor/opencor/files/14848962/jupyterlab.log));
> jupyternotebook: same as jupyterlab (see [log](https://github.com/opencor/opencor/files/14848958/jupyernotebook.log)); 

These are fixed with a small patch into `ipykernel` (ex. commit [0564aec](https://github.com/opencor/opencor/pull/2773/commits/0564aec8c455bb9cb16a7928f9d37a747fd8b255)) and a rebuild of the PythonPackages plugin.
 
> pythonshell: seems to be fine except for the fact that I couldn't use the Up and Down keys to recall previous commands. Instead, I got ^[[A and ^[[B, respectively (see [log](https://github.com/opencor/opencor/files/14848963/pythonshell.log)).

I'm aware of this -- it is because Python can't find a `readline` library; Linux has the same issue; Windows is fine. I'll look into why this is the case now that everything else is essentially working.

> Otherwise, as you might have seen, only the Windows debug CI passes (even though there are no new Windows debug binaries?!).

I've forgotten what the status of a Windows' debug version was although recall something about it being non-trivial because of library compatibility issues around Python, hence have yet to attempt a build.

>  All the other ones fail for some reasons.

It's the `PythonSupport` ones that fail -- I'll take a closer look.

**Comment by dbrnz** [comment]
@agarny adding the current working directory to `sys.path` is along the right lines for fixing the failing tests, it's just that what needs to be added is the directory containing the file/module being run. I'll sort it out.

**Comment by agarny** [comment]
> @agarny adding the current working directory to `sys.path` is along the right lines for fixing the failing tests, it's just that what needs to be added is the directory containing the file/module being run. I'll sort it out.

Yes, I realised that. I am just about to push something.

**Comment by dbrnz** [comment]
Hmm, I've done it with:
```

import pathlib

sys.argv = %1

sys.path.insert(0, pathlib.Path(sys.argv).parent)
```

**Comment by agarny** [comment]
> Hmm, I've done it with:
> 
> ```
> 
> import pathlib
> 
> sys.argv = %1
> 
> sys.path.insert(0, pathlib.Path(sys.argv).parent)
> ```

Doh! Of course, much easier! Ok, reverting my changes in favour of yours!

**Comment by dbrnz** [comment]
I see you've skipped over commit [55e8383](https://github.com/opencor/opencor/pull/2773/commits/55e8383cb9487a74e948f51c4463f0c7f8e4e51d) which is why the macOS builds are still failing...

**Comment by agarny** [comment]
> I see you've skipped over commit [55e8383](https://github.com/opencor/opencor/pull/2773/commits/55e8383cb9487a74e948f51c4463f0c7f8e4e51d) which is why the macOS builds are still failing...

Yes, I noticed that.

**Comment by dbrnz** [comment]
@agarny you're racing ahead! It needs to be:
```
sys.path.insert(0, str(pathlib.Path('%2').parent))
```
though, as well as we need:
```
sys.path.insert(0, '')
```
for the interactive case (NB. no `.` in the path, as vanilla Python has no `.`).

**Comment by agarny** [comment]
> @agarny you're racing ahead! It needs to be:
> 
> ```
> sys.path.insert(0, str(pathlib.Path('%2').parent))
> ```

Yes, I blindly added `sys.path.insert(0, pathlib.Path(sys.argv).parent)` to then realise that it wasn't working. I then realised about using `%2`, but I hadn't realised about the `str()`. Anyway, I have now got it to work locally and going to push things.

> though, as well as we need:
> 
> ```
> sys.path.insert(0, '')
> ```
> 
> for the interactive case (NB. no `.` in the path, as vanilla Python has no `.`).

Ok, I hadn't tested the interactive version. Have added it too.

**Comment by dbrnz** [comment]
The Windows issue is:

```
    sys.path.insert(0, str(pathlib.Path('C:\Users\dave\build\opencor\src\plugins\support\PythonSupport\tests\data\basictests.py').parent))
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: (unicode error) 'unicodeescape' codec can't decode bytes in position 2-3: truncated \UXXXXXXXX escape
```

which can be resolved by escaping backslashes passed in from Qt, since:
```
>>> str(pathlib.Path('C:\\Users\\dave\\build\\opencor\\src\\plugins\\support\\PythonSupport\\tests\\data\\basictests.py').parent)
'C:\\Users\\dave\\build\\opencor\\src\\plugins\\support\\PythonSupport\\tests\\data'
```

**Comment by agarny** [comment]
> The Windows issue is:
> 
> ```
>     sys.path.insert(0, str(pathlib.Path('C:\Users\dave\build\opencor\src\plugins\support\PythonSupport\tests\data\basictests.py').parent))
>                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> SyntaxError: (unicode error) 'unicodeescape' codec can't decode bytes in position 2-3: truncated \UXXXXXXXX escape
> ```
> 
> which can be resolved by escaping backslashes passed in from Qt, since:
> 
> ```
> >>> str(pathlib.Path('C:\\Users\\dave\\build\\opencor\\src\\plugins\\support\\PythonSupport\\tests\\data\\basictests.py').parent)
> 'C:\\Users\\dave\\build\\opencor\\src\\plugins\\support\\PythonSupport\\tests\\data'
> ```

Yeah, it makes sense indeed. Am going to look into it (from the Python side of things). This being said, why is the Windows debug version passing...!? Is that related to having no new Windows debug binaries?

**Comment by agarny** [comment]
> Thanks for this @dbrnz. It all compiles fine. However, besides my cleaning up and the few comments that I have made, I am having some issues with the Python scripts (I have only tested on macOS). For instance (note commit [4452777](https://github.com/opencor/opencor/commit/44527774e101642184679439e8197704bfaab5df) to prevent the debug warning):
> 
> * `ipython`: seems to be fine;
> * `jupyter`: seems to be fine;
> * `jupyterconsole`: I get various tracebacks (see [log](https://github.com/opencor/opencor/files/14848959/jupyterconsole.log));
> * `jupyterlab`: although it fires up JupyerLab fine and I could open the [lorenz.ipynb](https://github.com/opencor/opencor/blob/master/models/tests/jupyter/lorenz.ipynb) file, I got various tracebacks (see [log](https://github.com/opencor/opencor/files/14848962/jupyterlab.log));
> * `jupyternotebook`: same as `jupyterlab` (see [log](https://github.com/opencor/opencor/files/14848958/jupyernotebook.log)); and
> * `pythonshell`: seems to be fine except for the fact that I couldn't use the `Up` and `Down` keys to recall previous commands. Instead, I got `^[[A` and `^[[B`, respectively (see [log](https://github.com/opencor/opencor/files/14848963/pythonshell.log)).
> 
> Otherwise, as you might have seen, only the Windows debug CI passes (even though there are no new Windows debug binaries?!). All the other ones fail for some reasons.

I have just tried our various Python scripts one more time now that various changes have been done and:

- `ipython`: seems to be fine;
- `jupyter`: seems to be fine;
- `jupyterconsole`: seems to be fine;
- `jupyterlab`: seems to be fine;
- `jupyternotebook`: seems to be fine; and
- `pythonshell`: same as before (i.e. still an issue with the `Up` and `Down` keys) since this issue has yet to be fixed.

As mentioned earlier, I am going to check for the Windows release CI and escaping paths.

Otherwise, I would still like to understand what's going on with the Windows debug CI and why we have no new Windows debug binaries.

**Comment by dbrnz** [comment]
> This being said, why is the Windows debug version passing...!? Is that related to having no new Windows debug binaries?

I'm guessing so. It will be running Python 3.7 so there must be something in the (internal to Python) changes between 3.7 and 3.12 that has lead to the test failures.

I'm now in the process of building Windows debug binaries.

**Comment by agarny** [comment]
FTR, the Windows debug CI fails, it's just that for some reasons GitHub Actions doesn't report the failure as such. See the `Build and test OpenCOR` step at https://github.com/opencor/opencor/actions/runs/8547989492/job/23421036910.

**Comment by dbrnz** [comment]
Although I can build a Python debug executable under Windows which runs, it isn't able to load any of its extension modules (e.g. `_ssl`, `_ctypes`, etc), maybe because some flag hasn't been set at link time; and/or the libraries these modules use (e.g. `libffi`) haven't a debug version and need building from source; and/or the Python CMake build system needs updating to support Windows debug builds; and/or?? I recall similar difficulties with Python 3.7 -- at the time I had a go at tweaking the CMake build system, but can't remember the outcome -- reapplying these tweaks got the debug executable built, but no joy with the modules. Understanding why is likely going to require some low level debugging to see why `dlopen()` or similar is failing, not something I have the time/enthusiasm for anytime soon...

I see there is a comment [here](https://github.com/opencor/opencor/issues/1255#issuecomment-551298669) about disabling Jupyter and the Python console for Windows debug mode, but presumably a core Windows debug version does work under 3.7?

**Comment by agarny** [comment]
> Although I can build a Python debug executable under Windows which runs, it isn't able to load any of its extension modules (e.g. `_ssl`, `_ctypes`, etc), maybe because some flag hasn't been set at link time; and/or the libraries these modules use (e.g. `libffi`) haven't a debug version and need building from source; and/or the Python CMake build system needs updating to support Windows debug builds; and/or?? I recall similar difficulties with Python 3.7 -- at the time I had a go at tweaking the CMake build system, but can't remember the outcome -- reapplying these tweaks got the debug executable built, but no joy with the modules. Understanding why is likely going to require some low level debugging to see why `dlopen()` or similar is failing, not something I have the time/enthusiasm for anytime soon...
> 
> I see there is a comment [here](https://github.com/opencor/opencor/issues/1255#issuecomment-551298669) about disabling Jupyter and the Python console for Windows debug mode, but presumably a core Windows debug version does work under 3.7?

Yes, I wouldn't want you to waste time on the Windows debug version of Python. In fact, it was already the case with Python 3.7.5 (see [here](https://github.com/opencor/opencor/blob/2511941b2d0b22c041f46881d684aed3d1100234/CMakeLists.txt#L948-L951) and [here](https://github.com/opencor/opencor/blob/2511941b2d0b22c041f46881d684aed3d1100234/CMakeLists.txt#L957-L959)).

So, I am completely fine with things remaining the same with Python 3.12.2.

My understanding is that those Windows debug binaries will still allow to test things through our CI action, so I am fine with that.

**Comment by agarny** [comment]
Ok, at this stage, I think there are still two things that need to be finished/done:

1. Windows debug binaries with caveat that we can't use the `Python Console` window with them (i.e. same as with Python 3.7.5); and
2. Addressed the issue of the `Up` and `Down` keys in the `pythonshell`.

**Comment by dbrnz** [comment]
> Ok, at this stage, I think there are still two things that need to be finished/done:
>
> 1. Windows debug binaries with caveat that we can't use the Python Console window with them (i.e. same as with Python 3.7.5); and

This will need working out as to why DLLs for standard modules aren't recognised -- a rainy day activity in a few weeks time?

> 2. Addressed the issue of the Up and Down keys in the pythonshell.

This has been resolved in the last commits.

**Comment by agarny** [comment]
> > Ok, at this stage, I think there are still two things that need to be finished/done:
> > 1. Windows debug binaries with caveat that we can't use the Python Console window with them (i.e. same as with Python 3.7.5); and
> This will need working out as to why DLLs for standard modules aren't recognised -- a rainy day activity in a few weeks time?

I thought it was "just" a matter of building things for Windows debug and that, like for Python 3.7.5, the `Python Console` window couldn't be used in debug mode? Anyway, I know you have more important priorities, so I guess this will have to wait.

> > 2. Addressed the issue of the Up and Down keys in the pythonshell.
> This has been resolved in the last commits.

Has it really? I have just fetched your two commits and rebuilt everything from scratch on my Mac and I am still getting those `^[[A` and `^[[B`...?

**Comment by dbrnz** [comment]
>>> Addressed the issue of the Up and Down keys in the pythonshell.

>> This has been resolved in the last commits.

> Has it really? I have just fetched your two commits and rebuilt everything from scratch on my Mac and I am still getting those ^[[A and ^[[B...?

Oops! It's now fixed in the underlying `bin/python` -- I didn't check `pythonshell`... Soon sorted.

**Comment by dbrnz** [comment]
My last set of commits gets up/down arrows working and adds history to `pythonshell`. There's a new pre-built PythonPackages for macOS being uploaded, except GitHub appears to have network issues at the moment as it's very slooooow, so am giving up and will retry in the morning. The other platforms don't need new PythonPackages as their ones were never updated with the now reverted tweak to `ipykernel`.

**Comment by dbrnz** [comment]
The macOS PythonPackages binary has now uploaded so everything should be good to go (apart from Windows debug mode).

I'm not sure why the PythonSupport tests have started failing. Manually running `basictests.py` (under both macOS and Ubuntu 20) finds that the test output is missing the following -- has something changed wrt COMBINE??

```
---------------------------------------------------------------------
            Unsupported local COMBINE archive (algorithm)
---------------------------------------------------------------------
 - Open simulation
 - Check simulation:
    - Valid: no
    - Issues:
       - Information: unsupported algorithm (KISAO:0000029).

---------------------------------------------------------------------
       Unsupported local COMBINE archive (algorithm parameter)
---------------------------------------------------------------------
 - Open simulation
 - Check simulation:
    - Valid: no
    - Issues:
       - Information: unsupported algorithm parameter (KISAO:0000219).

---------------------------------------------------------------------
           Invalid local COMBINE archive (parameter value)
---------------------------------------------------------------------
 - Open simulation
 - Check simulation:
    - Valid: no
    - Issues:
       - Error: the value of 'Absolute tolerance' (KISAO:0000211) must be a number greater or equal to zero.
       - Error: the value of 'Integration method' (KISAO:0000475) must be 'Adams-Moulton' or 'BDF'.
       - Error: the value of 'Interpolate solution' (KISAO:0000481) must be 'true' or 'false'.
       - Error: the value of 'Iteration type' (KISAO:0000476) must be 'Functional' or 'Newton'.
       - Error: the value of 'Linear solver' (KISAO:0000477) must be 'Dense', 'Banded', 'Diagonal', 'GMRES', 'BiCGStab' or 'TFQMR'.
       - Error: the value of 'Lower half-bandwidth' (KISAO:0000480) must be an integer greater or equal to zero.
       - Error: the value of 'Maximum number of steps' (KISAO:0000415) must be an integer greater than zero.
       - Error: the value of 'Maximum step' (KISAO:0000467) must be a number greater or equal to zero.
       - Error: the value of 'Preconditioner' (KISAO:0000478) must be 'None' or 'Banded'.
       - Error: the value of 'Relative tolerance' (KISAO:0000209) must be a number greater or equal to zero.
       - Error: the value of 'Upper half-bandwidth' (KISAO:0000479) must be an integer greater or equal to zero.
```

**Comment by dbrnz** [comment]
> I'm not sure why the PythonSupport tests have started failing.

The tests were in fact succeeding; however not all of the `pythonshell` output was being captured; a `flush()` has fixed things.

**Comment by dbrnz** [comment]
Finally all tests are passing!

The Windows debug failure was due to `escapedPath()` -- I suspect its result was left on the stack, with the unused part (after a return) being cleaned up when running in debug mode.

**Comment by dbrnz** [comment]
Nothing more from me -- let's merge!

**Comment by dbrnz** [comment]
> Strange about escapedPath() since something similar was also used before https://github.com/opencor/opencor/commit/7266c5c794c08df51c91c98a66eb3c9be487927a).

That though wasn't calling a separate function and playing with `wchar_t *` and `std::wstring` but instead passing a `QString` (with Qt's reference counting).

**Comment by agarny** [comment]
> Nothing more from me -- let's merge!

Ok, will have a final reivew sometime today and, everything going well, I will merge it be the end of the day, and will look into generating a new snapshot.

> That though wasn't calling a separate function and playing with `wchar_t *` and `std::wstring` but instead passing a `QString` (with Qt's reference counting).

I would have thought that it would have been fine since `escapedPath()` returns a `std::string()` but maybe I should have used `return std::string((const wchar_t *) ePath.utf16());` rather than `return (const wchar_t *) ePath.utf16();`? Anyway, it now all works fine and tat's what matters!

**Pull request merged**